# Laser Off Code

In [ ]:
############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

# Import

In [1]:
import time
from time import sleep, monotonic
import datetime
import numpy as np
import matplotlib.pyplot as plt
import sys
import pyvisa
import qcodes as qc
from qcodes.dataset import Measurement
from qcodes.dataset import do0d
from qcodes.dataset.experiment_container import new_experiment, load_experiment_by_name
from qcodes.dataset.plotting import plot_by_id
from qcodes.dataset.data_set import load_by_id, load_by_counter
from qcodes import initialise_or_create_database_at, new_data_set, new_experiment
from qcodes.station import Station
initialise_or_create_database_at("./2026-04-17_SNSPD5.db")
from functions import quick_check
from functions import calibrate
import snspd
params = snspd.snspd('D:\SNSPD\SNSPD2\SNSPD5\snspd5-1.yaml')

# Set up experiment
exp_name = 'SNSPD5'
sample_name = '00'

try:
    exp = qc.load_experiment_by_name(exp_name, sample=sample_name)
    print('Experiment loaded. Last ID no:', exp.last_counter)
except ValueError:
    exp = new_experiment(exp_name, sample_name)
    print('Started new experiment')

<>:20: SyntaxWarning: invalid escape sequence '\S'
<>:20: SyntaxWarning: invalid escape sequence '\S'
C:\Users\QNL\AppData\Local\Temp\ipykernel_15884\159542524.py:20: SyntaxWarning: invalid escape sequence '\S'
  params = snspd.snspd('D:\SNSPD\SNSPD2\SNSPD5\snspd5-1.yaml')


Logging hadn't been started.
Activating auto-logging. Current session state plus future input saved.
Filename       : C:\Users\QNL\.qcodes\logs\command_history.log
Mode           : append
Output logging : True
Raw input log  : False
Timestamping   : True
State          : active
Qcodes Logfile : C:\Users\QNL\.qcodes\logs\260420-15884-qcodes.log
Experiment loaded. Last ID no: 12


In [6]:
station = Station(config_file="friesland.yaml")

In [7]:
laser = station.load_instrument("laser", revive_instance=True)

In [8]:
pm100usb = station.load_instrument("pm100usb", revive_instance=True)

In [9]:
pms120 = station.load_instrument("pms120", revive_instance=True)

# Calibration

Checking single fibre from fridge fibre port 1 to fibre port 2. 

In [6]:
params.connection1

'single fibre'

In [16]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection1)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}\n")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-17 16:53:01,300 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 1. 
1
Power on PM100USB after attenuator is:0.00213529891
Power on PM120 after 90% beamsplitter port is:0.00483268499
Attenuation is:-6.5974389124266075

Laser enable status: False


In [17]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection1)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}\n"")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-17 16:54:21,791 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 2. 
2
Power on PM100USB after attenuator is:0.00212446461
Power on PM120 after 90% beamsplitter port is:0.00480654463

Laser enable status: False


In [19]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection1)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-17 16:55:13,653 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 3. 
3
Power on PM100USB after attenuator is:0.00212647882
Power on PM120 after 90% beamsplitter port is:0.00481490931
Attenuation is:-6.595466501707738
Laser enable status: False


In [22]:
plaser = power90/params.bs90 
pin = params.bs10*plaser

In [24]:
plaser

0.005280618074468063

In [23]:
pin

0.0004657087644680629

In [25]:
laser.power()

7.0

In [27]:
pm100usb.

2026-04-17 17:11:44,295 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



'1\n'

In [28]:
############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: False


Wrapped fibre comi ng oit of fridge and joining fridge ports in foil

In [29]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection1)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-17 17:34:36,475 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 4. 
4
Power on PM100USB after attenuator is:0.00200799364
Power on PM120 after 90% beamsplitter port is:0.00482118316
Attenuation is:-6.340824022017312
Laser enable status: False


REstart kernel

In [9]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection1)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-17 17:35:40,277 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 5. 
5
Power on PM100USB after attenuator is:0.00199563126
Power on PM120 after 90% beamsplitter port is:0.0047992249
Attenuation is:-6.333828962233381
Laser enable status: False


In [10]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

print(pms120.power())
print(pm100usb.power())

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-17 17:37:44,899 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



0.00480131619
0.00199174206
Laser enable status: False


In [11]:
############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: False


Other power meter connected to 10% output

In [12]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

print(pms120.power())
print(pm100usb.power())

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-17 17:44:06,841 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



0.000383867678
1.99510445e-14
Laser enable status: False


In [13]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection1)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-17 17:45:15,437 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 6. 
6
Power on PM100USB after attenuator is:1.42649972e-11
Power on PM120 after 90% beamsplitter port is:0.000382783415
Attenuation is:64.14207721703565
Laser enable status: False


Power on PM120 after 90% beamsplitter port is:0.0047992249

Connection 2 - pm120 is in the position downstream of  10% 

In [14]:
params.connection2

'ferrule to ferrule'

In [15]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection2)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-17 18:12:33,746 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 7. 
7
Power on PM100USB after attenuator is:1.99510445e-14
Power on PM120 after 90% beamsplitter port is:0.000333292322
Attenuation is:92.08385971482313
Laser enable status: False


In [16]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection2)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-17 18:12:58,335 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 8. 
8
Power on PM100USB after attenuator is:1.99510445e-14
Power on PM120 after 90% beamsplitter port is:0.000332687632
Attenuation is:92.07597318307839
Laser enable status: False


In [17]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection2)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-17 18:13:47,414 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 9. 
9
Power on PM100USB after attenuator is:7.14247411e-12
Power on PM120 after 90% beamsplitter port is:0.000332322757
Attenuation is:66.53237707127558
Laser enable status: False


# 4K Measurement

In [2]:
params.connection1

'single fibre'

In [8]:
params.bs10

0.08819209378534265

In [9]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection2)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-20 10:27:57,766 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 10. 
10
Power on PM100USB after attenuator is:7.14247411e-12
Power on PM120 after 90% beamsplitter port is:1.67331768e-06
Attenuation is:43.552560247118436
Laser enable status: False


In [10]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection2)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-20 10:29:56,371 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 11. 
11
Power on PM100USB after attenuator is:7.14247411e-12
Power on PM120 after 90% beamsplitter port is:1.64079881e-06
Attenuation is:43.46732957334219
Laser enable status: False


In [11]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)

power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.connection2)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-20 10:31:57,318 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 12. 
12
Power on PM100USB after attenuator is:7.14247411e-12
Power on PM120 after 90% beamsplitter port is:1.64919106e-06
Attenuation is:43.48948596843721
Laser enable status: False


Measuring beam splitter with PM120 to check that the powermeter is working

In [10]:
params.position1

'PM120_bs90'

In [12]:
############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: False


PM120 connected to 90% BS port. PM100USB disconnected. 10% port connected to fridge in Connection 1. 

In [13]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)


power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.position1)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-20 11:03:56,474 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 13. 
13
Power on PM100USB after attenuator is:7.14247411e-12
Power on PM120 after 90% beamsplitter port is:0.00480654463
Attenuation is:78.13510604151189
Laser enable status: False


In [14]:
############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: False


PM120 connected to 100% BS port. PM100USB disconnected. 90% port disconnected.

In [16]:
############################ TURN LASER ON ############################ 
laser.enable(True)
print(f'Laser enable status: {laser.enable()}')
time.sleep(10)


power10, power90, attenuation = quick_check(bs10=params.bs10, bs90=params.bs90, pmeter10=pm100usb, pmeter90=pms120, attenuator_name = params.position2)

print(f"Power on PM100USB after attenuator is:{power10}\nPower on PM120 after 90% beamsplitter port is:{power90}\nAttenuation is:{attenuation}")

############################ TURN LASER OFF ############################ 
laser.enable(False)
print(f'Laser enable status: {laser.enable()}')

Laser enable status: True


2026-04-20 11:09:16,009 ¦ py.warnings ¦ WARNING ¦ warnings ¦ _showwarnmsg ¦ 110 ¦ C:\Users\QNL\anaconda3\envs\qcodes\Lib\site-packages\pyvisa\resources\messagebased.py:702: UserWarning: read string doesn't end with termination characters
  return self.read()



Starting experimental run with id: 15. 
15
Power on PM100USB after attenuator is:1.99510445e-14
Power on PM120 after 90% beamsplitter port is:0.000469726278
Attenuation is:93.57405519170712
Laser enable status: False
